tstest/integration/vms/vms_test.go

```diff
@@ -0,0 +1,457 @@
+	})
+
+	close(cancels)
+	for cancel := range cancels {
+		//lint:ignore SA9001 They do actually get ran
+		defer cancel()
+
+		t.Logf("%d vms in the pipe", len(cancels))
+
+		if len(cancels) == 0 {
+			t.Log("all vms started")
+			break
+		}
+	}
+
+	t.Log("waiting for vms to register")
+	waiter := time.NewTicker(10 * time.Second)
+	defer waiter.Stop()
+	n := 0
+	for {
+		<-waiter.C
+		ipLock.Lock()
+		if len(ipMap) == len(distros) {
+			ipLock.Unlock()
+			break
+		} else {
+			if n%5 == 0 {
+				t.Logf("ipMap:   %d", len(ipMap))
+				t.Logf("distros: %d", len(distros))
+			}
+		}
+		n++
+		ipLock.Unlock()
+	}
+
+	ipLock.Lock()
+	defer ipLock.Unlock()
+	t.Run("join-net", func(t *testing.T) {
+		for port := range ipMap {
+			port := port
+			t.Run(fmt.Sprint(port), func(t *testing.T) {
+				t.Parallel()
+				t.Logf("about to ssh into 127.0.0.1:%s", port)
+				timeout := 5 * time.Minute
+
+				sshcmd := fmt.Sprintf("ssh -o StrictHostKeyChecking=no -o GlobalKnownHostsFile=/dev/null -o UserKnownHostsFile=/dev/null -i %s ts@127.0.0.1 -p %s", filepath.Join(dir, "machinekey"), port)
+
+				cmd := exec.Command("sh", "-c", sshcmd+" -- uname -av")
+				t.Logf("running: %s -- uname -av", sshcmd)
+				cmd.Stdout = os.Stdout
+				cmd.Stderr = os.Stderr
+				err := cmd.Run()
+				if err != nil {
+					t.Fatalf("can't get uname -av output: %v", err)
+				}
+
+				e, _, err := expect.Spawn(sshcmd, timeout, expect.Verbose(true), expect.VerboseWriter(os.Stdout))
+				if err != nil {
+					t.Fatalf("%s: can't register a shell session: %v", port, err)
+				}
+				defer e.Close()
+
+				_, _, err = e.Expect(regexp.MustCompile(`(\$|\>)`), timeout)
+				if err != nil {
+					t.Fatalf("%s: can't get a shell: %v", port, err)
+				}
+				t.Logf("got shell for %s", port)
+				err = e.Send(fmt.Sprintf("sudo tailscale up --login-server %s\n", loginServer))
+				if err != nil {
+					t.Fatalf("%s: can't send tailscale up command: %v", port, err)
+				}
+				_, _, err = e.Expect(regexp.MustCompile(`Success.`), timeout)
+				if err != nil {
+					t.Fatalf("can't extract URL: %v", err)
+				}
+			})
+		}
+	})
+
+	if numNodes := cs.NumNodes(); numNodes != len(ipMap) {
+		t.Errorf("wanted %d nodes, got: %d", len(ipMap), numNodes)
+	}
+}
+
+const metaDataTemplate = `instance-id: {{.ID}}
+local-hostname: {{.Hostname}}`
+
+const userDataTemplate = `#cloud-config
+#vim:syntax=yaml
+
+cloud_config_modules:
+ - runcmd
+
+cloud_final_modules:
+ - [users-groups, always]
+ - [scripts-user, once-per-instance]
+
+users:
+  - name: ts
+    plain_text_passwd: hunter2
+    groups: [ wheel ]
+    sudo: [ "ALL=(ALL) NOPASSWD:ALL" ]
+    shell: /bin/sh
+    ssh-authorized-keys:
+    - {{.SSHKey}}
+    - ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIC1e4qhGYEUCNoCYHUqfvPSkBfVdlIjmwQI7q8eibeWw
+
+write_files:
+  - path: /etc/cloud/cloud.cfg.d/80_disable_network_after_firstboot.cfg
+    content: |
+      # Disable network configuration after first boot
+      network:
+        config: disabled
+
+runcmd:
```

Comment:
let's dynamically generate this list based on new methods on Distro, so we run the minimal set of commands necessary for each image